src/template_finder.py

```diff
@@ -9,235 +9,222 @@
 import time
 import os
 from config import Config
-from utils.misc import cut_roi, load_template, list_files_in_folder, alpha_to_mask, roi_center, color_filter
+from utils.misc import cut_roi, load_template, list_files_in_folder, alpha_to_mask, roi_center, color_filter, mask_by_roi
+from functools import cache
 
 template_finder_lock = threading.Lock()
 
+@dataclass
+class Template:
+    name: str = None
+    img_bgra: np.ndarray = None
+    img_bgr: np.ndarray = None
+    img_gray: np.ndarray = None
+    alpha_mask: np.ndarray = None
+
 @dataclass
 class TemplateMatch:
     name: str = None
     score: float = -1.0
-    center: tuple[float, float] = None
+    center: tuple[int, int] = None
+    center_monitor: tuple[int, int] = None
     region: list[float] = None
+    region_monitor: list[float] = None
     valid: bool = False
 
-class TemplateFinder:
+
+TEMPLATE_PATHS = [
+    "assets\\templates",
+    "assets\\npc",
+    "assets\\shop",
+    "assets\\item_properties",
+    "assets\\chests",
+    "assets\\gamble",
+]
+
+@cache
+def _templates() -> dict[Template]:
+    paths = []
+    templates = {}
+    for path in TEMPLATE_PATHS:
+        paths += list_files_in_folder(path)
+    for file_path in paths:
+        file_name: str = os.path.basename(file_path)
+        if file_name.lower().endswith('.png'):
+            key = file_name[:-4].upper()
+            template_img = load_template(file_path)
+            templates[key] = Template(
+                name = key,
+                img_bgra = template_img,
+                img_bgr = cv2.cvtColor(template_img, cv2.COLOR_BGRA2BGR),
+                img_gray = cv2.cvtColor(template_img, cv2.COLOR_BGRA2GRAY),
+                alpha_mask = alpha_to_mask(template_img)
+            )
+    return templates
+
+def get_template(key):
+    return _templates()[key].img_bgr
+
+def _process_template_refs(ref: Union[str, np.ndarray, list[str]]) -> list[Template]:
+    templates = []
+    if type(ref) != list:
+        ref = [ref]
+    for i in ref:
+        # if the reference is a string, then it's a reference to a named template asset
+        if type(i) == str:
+            templates.append(_templates()[i.upper()])
+        # if the reference is an image, append new Template class object
+        elif type(i) == np.ndarray:
+            templates.append(Template(
+                img_bgr = i,
+                img_gray = cv2.cvtColor(i, cv2.COLOR_BGR2GRAY),
+                alpha_mask = alpha_to_mask(i)                    
+            ))
+    return templates
+
+def _grayscale_template_and_image(template: Template, img: np.ndarray) -> tuple[np.ndarray]:
+    return template.img_gray, cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+
+def _color_filter_template_and_image(template: Template, img: np.ndarray, color_match: list = None) -> tuple[np.ndarray]:
+    return color_filter(template.img_bgr, color_match)[1], color_filter(img, color_match)[1]
```

Comment:
I really like the functional approach taken. But for these two functions that are used once in a pretty obvious if statement, I dont see the benefit.